File: src/2815.max-pair-sum-in-an-array.2902/solution.ts
function getMaxDigit (n:number) {
    let res = 0;
    while (n > 0) {
        const digit = n % 10;
        res = Math.max(res, digit);
        n = (n - digit) / 10;
    }
    return res;
}

function maxSum (nums: number[]): number {
    let result = -1;
    const maxByDigit = new Array<number>(10).fill(0);
    for (const n of nums) {
        const maxDigit = getMaxDigit(n);
        if (maxByDigit[maxDigit] !== 0) {
            result = Math.max(result, n + maxByDigit[maxDigit]);
        }
        maxByDigit[maxDigit] = Math.max(n, maxByDigit[maxDigit]);
    }

    return result;
}

File: src/2815.max-pair-sum-in-an-array.2902/solution.java
class Solution {
    public int maxSum(int[] nums) {
        int[] maxs = new int[10];
        int[] nexts = new int[10];
        for(int num : nums){
            int index = calc(num);
            if(maxs[index]<=num){
                nexts[index] = maxs[index];
                maxs[index] = num;
            }else if(nexts[index]<num){
                nexts[index] = num;
            }
        }
        int result = -1;
        for(int i=0;i<10;i++){
            if(nexts[i] != 0){
                result = Math.max(result,maxs[i]+nexts[i]);
            }
        }
        return result;
    }

    private int calc(int num){
        int result = 0;
        while (num != 0){
            result = Math.max(result,num%10);
            num /= 10;
        }
        return result;
    }
}